### DISCUSSION TITLE
Master on cloud and agents behind NAT

### DISCUSSION THREAD

**Discussion by tdtgit** [k3s-io/k3s#9908] [Q&A]
Hi guys,

First, thanks for the good software made self-hosted cluster possible. I already able to setup my K3S cluster by Tailscale provider. But I assume I can do better with it:

- Tailscale seems to be slower and CPU intense compare to WireGuard? I'm planning to host some intensive media traffic between the nodes so the CPU resources later will be the problem.
- Vendor lock-in. I know we can try Headscale, but I want to try WireGuard first.

Please help to advise the `wireguard-native` backend setup with my below servers, some params I don't know what to put in. Base on this doc: https://docs.k3s.io/networking/distributed-multicloud#embedded-k3s-multicloud-solution, the command would be:

* Master node (public cloud, have public IP):
  * Public IP: 103.70.xx.xx
  * `--node-external-ip=103.70.xx.xx --flannel-backend=wireguard-native --flannel-external-ip`
* Agent node (home network, behind NAT, private IP only):
  * Public IP (of router) 1.52.xx.xx, Private IP: 192.168.9.xx
  * `--node-external-ip=<AGENT_EXTERNAL_IP>`: what is correct value of `AGENT_EXTERNAL_IP` to put in? I don't think we should use `1.52.xx.xx`

Another problem I want connect to my cameras in my home network, placed at `192.168.9.xx`. That's somehow impossible with Tailscale setup right now, even with `extraArgs`.

**Comment by ostapkostyk** [comment]
Hello,

I'm facing almost the same issue and also looking for an answer.
But I have HA setup with tree master nodes with public IPs and embedded ETCD. They configured with ```--flannel-backend=wireguard-native --flannel-external-ip``` options as well.
In internet I found following solution:
- install the server with following options: ```--node-external-ip <PUBLIC_IP>  --advertise-address <WIREGUARD_IP_NODE1> --flannel-iface wg0```
- install an agent with following configuration:  
```curl -sfL https://get.k3s.io | \
   K3S_URL=https://<WIREGUARD_IP_NODE1>:6443 \ 
   K3S_TOKEN="<replaced>" \
   sh -s - --node-label node=02 \
    --node-ip <WIREGUARD_IP_NODE2> \
    --flannel-iface wg0
  ```

I have not tried it yet and would like to ask gurus here will it work. @tdtgit you may try it if you are willing to experiment.

In my case, I'd also like to ask a community can I re-configure masters without interrupting the work of embedded ETCD as I afraid to kill the whole cluster.

**Reply by tdtgit** [reply]
Hi @ostapkostyk , thanks for your input. I have spent some hours to try (slowly) to understand how `wireguard-native` options work on this case and yes, I managed to have a working clusters with it. So the keys here is:

* You don't need to install or config Wireguard. That's what I do at first 🤪  Actually k3s will managed the Wireguard installation, create a network `flannel-wg` and join the nodes by themself.
* You have to connect the agents to master through <`public IP address`>:<6443> at first. Guessing they're exchange the node information here to create a Wireguard network
  * After the agents joined the cluster, you can change `K3S_URL` to server's Wireguard IP. But I'm not sure if the agents change the public IP or private IP, they have to exchange the information again by the server's public IP.
  * So your server have to open the control plane to the public internet (at least in my case, since the agents have dynamic IPs and the provider not provide any ways to update the firewall rules by API)
  * 
--------------------
Command ran by the server and agents:

Server:
```
curl -sfL https://get.k3s.io | sh -s - --node-external-ip=103.70.xx.xx --flannel-backend=wireguard-native --disable servicelb  --disable traefik --disable metrics-server --disable local-storage
```

Agents:
```
K3S_URL=https://103.70.xx.xx:6443 K3S_TOKEN=<REDACTED> k3s agent
```
The connection between the nodes has been verified by `iftop`, all the connections is through `flannel-wg` interface, except `:6443` as I explained.

Cheers.

--------------------
`wg show` output. FYI.
```
interface: flannel-wg
  public key: <REDACTED, not related>
  private key: (hidden)
  listening port: 51820

peer: <REDACTED, not related>
  endpoint: <My home public IP>:51820
  allowed ips: 10.42.1.0/24
  latest handshake: 56 seconds ago
  transfer: 5.50 MiB received, 8.79 MiB sent
  persistent keepalive: every 25 seconds

peer: <REDACTED, not related>
  endpoint: <My office public IP>:51820
  allowed ips: 10.42.2.0/24
  latest handshake: 1 minute, 43 seconds ago
  transfer: 2.67 MiB received, 3.12 MiB sent
  persistent keepalive: every 25 seconds
```
